### PR TITLE
[Core] Update to new exception hierarchy.

### LINKED ISSUE #74: Update BAL to use new exception hierarchy

### PR DESCRIPTION
**Pull request by elliotcmorris**
Closes #74

Migrate bal errors to new exceptions in openassetio errors namespace.

Needs work in progress (https://github.com/OpenAssetIO/OpenAssetIO/issues/1073) to be merged

### CODE REVIEW

File: plugin/openassetio_manager_bal/bal.py [hunk 1]
```diff
@@ -330,6 +330,17 @@ def _copy_and_expand_trait_properties(entity_version_dict: dict, library: dict)
...
+    """
+    An exception raised for a reference to an entity reference that is
+    formatted incorrectly or otherwise malformed according to BALs
+    expectations.
+    """
+
+    def __init__(self, malformed_reason: str, entity_ref: str):
+        super().__init__(f"{malformed_reason} ({entity_ref})")
+
+
```
Comment (foundrytom) [inline plugin/openassetio_manager_bal/bal.py:343]: [consider] `bal.py` is meant to represent some existing system, as such, it's not meant to know about OpenAssetIO concepts, with the bridge happening in the Interface class - might be better to define this there instead?

Comment (elliotcmorris) [reply plugin/openassetio_manager_bal/bal.py:343]: To my mind this MalformedEntityReference now is about an existing system, as it's a placeholder for an exception about an entityReference not living up to BALs expectations.
You think a rename might help to to bring it in line with the other exceptions defined in this file, like `MalformedBALReference`?

Comment (foundrytom) [reply plugin/openassetio_manager_bal/bal.py:343]: BAL itself doesn't have any expectations on the string though, i _only_ traffics the `EntityInfo` class - that defines its data, the conversion to/from an Entity Reference is done in the interface layer as part of the "bridging to OpenAssetIO concepts"

Comment (foundrytom) [reply plugin/openassetio_manager_bal/bal.py:343]: (This is why `_build_entity_ref` and `_parse_entity_ref` are not in `bal.py` - to help enforce the concept of the `BasicAssetLibraryInterface` class being where you bridge your two systems together. eg `bal.py` is logically some closed source independent tool.

**Comment by feltech** [comment]
Would be good to retarget CI to https://github.com/OpenAssetIO/OpenAssetIO/pull/1109 with a `drop!` commit, adding a little bit of test coverage for that PR.

**Comment by elliotcmorris** [comment]
> Would be good to retarget CI to [OpenAssetIO/OpenAssetIO#1109](https://github.com/OpenAssetIO/OpenAssetIO/pull/1109) with a `drop!` commit, adding a little bit of test coverage for that PR.

Good idea, did that.

File: plugin/openassetio_manager_bal/bal.py (hunk 1 above)
Comment (elliotcmorris) [reply plugin/openassetio_manager_bal/bal.py]: [05f5109](https://github.com/OpenAssetIO/OpenAssetIO-Manager-BAL/pull/75/commits/05f5109db36e7a0ec73ee9a665a6701b21066407)

**Comment by elliotcmorris** [comment]
OpenAssetIO change merged, dropped the retarget commit.

**Comment by elliotcmorris** [comment]
Shouldn't have dropped the retarget commit ... BAL uses a release natively, added a retarget commit to OpenAssetio/main this time.

**Review by feltech — APPROVED** [review]

**Pull request merged**